envoy/network/filter.h

```diff
@@ -400,6 +404,85 @@ class ListenerFilterManager {
  */
 using ListenerFilterFactoryCb = std::function<void(ListenerFilterManager& filter_manager)>;
 
+/**
+ * QUIC Listener Filter
+ */
+class QuicListenerFilter {
+public:
+  virtual ~QuicListenerFilter() = default;
+
+  /**
+   * Called when a new connection is accepted, but before a Connection is created.
+   * Filter chain iteration can be stopped if the cb shouldn't be accessed any more
+   * by returning `FilterStatus::StopIteration`, or continue the filter chain iteration
+   * by returning `FilterStatus::ContinueIteration`. Reject the connection by closing
+   * the socket and returning `FilterStatus::StopIteration`.
+   * @param cb the callbacks the filter instance can use to communicate with the filter chain.
+   * @param server_preferred_addresses the server's preferred addresses to be advertised.
+   * @return status used by the filter manager to manage further filter iteration.
+   */
+  virtual FilterStatus onAccept(ListenerFilterCallbacks& cb) PURE;
+
+  /**
+   * Called before connection creation.
+   * @return false if the given preferred address is incomplatible with this filter and the listener
+   * shouldn't advertise the given preferred address. I.e. onAccept() would have behaved differently
+   * if the connection socket's destination address were the preferred address.
+   */
+  virtual bool isCompatibleWithServerPreferredAddress(
```

Comment:
I have mixed feelings about using the QUIC address class here.  on the one hand, we already have a parsed address passed down from the QUIC stack and it's faster to use it for folks who don't want to use Envoy utils. On the other hand I don't like QUIC APIs leaking into the envoy API section.  I'm inclined to say it's worth converting these to Envoy addresses, but if @RyanTheOptimist thinks it's better this way maybe he could suggest a !google tiebreaker?